### DISCUSSION TITLE
Best Practice For Storing an array of data in entity component

### DISCUSSION THREAD

**Discussion by simeonradivoev** [sebas77/Svelto.ECS#82] [Q&A] [answered]
I have the use case, where an entity component must hold a navigation path. Currently, I use `NativeDynamicArray` and that does the trick but is it the best practice since that memory would be all over the place. That also means I have to manually dispose of the array when an entity is removed. Unity's DOTS has the `Dynamic Buffers` is there an alternative here?

**Comment by sebas77** [comment, marked as answer]
Using NativeDynamicArray for this kind of thing is fine. You don't need necessarily to use entities for this though, you can have your own datastructure, in your own class, that you inject in engines. This practice is mixing OOP and ECS and I discussed It in my article about OOP abstraction in an ECS centric application.
You can also use a managed component, implementors can be any kind of object.
Since in future I want to make implementors less and less useful, I will introduce the concept of IManagedEntityComponent, so you can use normal c# arrays. Of course, you cannot burstify them, so if you need burst, either your own data structure or NativeDynamicArray is ok.

**Comment by simeonradivoev** [comment]
I implemented my own buffer that resembles the unity buffer logic. That is burstable and also in continuous memory. It works by having an array that holds fixed capacity buckets and then is unsafely referenced using pointers and the unity `UnsafeList` type. It is blazing fast since it continues memory. The only downside is they need to be declared and cannot grow beyond the initial capacity (similar to the unity buffers). The API for registering is something like this: 
`````````` C#
map.CreateBufferCollection<PathNode>(group,capicity,bufferCapcity);
```````````
and retrieving it is as simple as:
`````````` C#
map.TryGetBufferCollection<Node>(group, out var bufferCollection);
bufferCollection.TryGetBuffer(entityId, out var buffer);
``````````````````
They are also managed by a reactive engine that removes buffers when an entry is removed.
I might share the code at some point but it's for an in-house project and safety checks are sparse and the API isn't that robust.

There are also other possibilities:
* There is also another option I found out. Unity provides a `FixedList`, `FixedString`, and `FixedBytes` structs that can be used.
* You can make use of [fixed-size buffers](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code#fixed-size-buffers), they work wonderfully but are a bit limited . You can probably also make a fixed list struct with them as long as you are fine with unsafe code.

**Reply by sebas77** [reply]
Svelto provides native data structures for native arrays that are resizable too. Not sure if you missed that.

**Reply by simeonradivoev** [reply]
what type is that ?

**Reply by sebas77** [reply]
I had to re read everything. Did You implemente a way to bind svelto Nativedynamicarray to entities? did I get it right? If you didn't use nativedynamicarrays I don't see why you shouldn't. It seems an interesting extension to publish, what do you think?

**Reply by simeonradivoev** [reply]
I did not use NativeDynamicArray. I wanted all group data to be in linear memory since I needed to handle tens of thousands of entities with complex logic and NativeDynamicArrays didn't cut it.

**Reply by simeonradivoev** [reply]
I basically allocate a huge chunk of memory and give each entity a slice of the memory.

**Reply by sebas77** [reply]
What makes you think that data stored in the native dynamic array is not continuous in memory?

I should implement a slice datastructrure, I'll probably do it next release

Anyway using a linked list of big buckets could be a good strategy too depending the needs

**Reply by sebas77** [reply]
Feel free to share the code I am sure the community will appreciate it!
